[DOCS] Clarify testing section

ChrisChinchilla opened the pull request:
Closes #5166 this PR attempts to make the tests section of the `contributing.rst` doc. There are a handful of outstanding questions to make this ready for review as comments below.

### Checklist
- [ ] Code compiles correctly
- [ ] All tests are passing
- [ ] New tests have been created which fail without the change (if possible)
- [ ] README / documentation was extended, if necessary
- [ ] Changelog entry (if change is visible to the user)
- [ ] Used meaningful commit messages

ChrisChinchilla commented:
Question 1

Is there a way to pass `--no-smt` via `tests.sh` to the underlying `soltest.sh`?

ChrisChinchilla commented:
Question 2

How can I tell which tests need evmone and libz3?

leonardoalt commented:
1) `tests.sh` seems to use env variable `$SMT_FLAGS` which could be used for that I guess, even though I searched and it doesn't seem to be used anywhere at the moment.

2) 
- latest z3: `libsolidity/smtCheckerTests` and `libsolidity/smtCheckerTestsJSON`
- `evmone`:  `libsolididty/semanticTests`, `libsolidity/GasCosts`, `libsolidity/SolidityEndToEndTest` I think, there might be more.

ChrisChinchilla commented:
@leonardoalt I'm not sure I understand your comment?

leonardoalt commented:
Answering to your questions 1) and 2) in the previous comments

leonardoalt commented:
@ChrisChinchilla any updates here?

ChrisChinchilla commented:
Slipped my mind sorry @leonardoalt 

Your answers will still need a bit more work, but will do now

stackenbotten commented:
There was an error when running `chk_coding_style` for commit `c475176fd20df2554bf07d80f9faa2f586e0cdf2`:
```
Error: Trailing whitespace found:
 docs/contributing/tests-list.csv:1:M-oM-;M-?Test suite,Tests^M
 docs/contributing/tests-list.csv:2:**Command line**,Passing unknown options to CLI^M
 docs/contributing/tests-list.csv:3:,Passing files that are not found^M
 docs/contributing/tests-list.csv:4:,Passing files that are not files^M
 docs/contributing/tests-list.csv:5:,Passing empty remappings^M
 docs/contributing/tests-list.csv:6:,General tests (`Full details <https://github.com/ethereum/solidity/tree/develop/test/cmdlineTests>`_)^M
 docs/contributing/tests-list.csv:7:**Compilation**,Compiling other contracts and libraries^M
 docs/contributing/tests-list.csv:8:,MultiSigWallet^M
 docs/contributing/tests-list.csv:9:,Corion^M
 docs/contributing/tests-list.csv:10:,Gnosis^M
 docs/contributing/tests-list.csv:11:,Milestone tracker^M
 docs/contributing/tests-list.csv:12:,String Utils^M
 docs/contributing/tests-list.csv:13:,Documentation examples^M
 docs/contributing/tests-list.csv:14:,Library checksum^M
 docs/contributing/tests-list.csv:15:,Long library names^M
 docs/contributing/tests-list.csv:16:,Linking libraries^M
 docs/contributing/tests-list.csv:17:,Overwriting already compiled files^M
 docs/contributing/tests-list.csv:18:,"Assemble, yul, strict-assembly and optimize compiled outputs"^M
 docs/contributing/tests-list.csv:19:,Standard input to compiler^M
 docs/contributing/tests-list.csv:20:,soljson via the fuzzer^M
 docs/contributing/tests-list.csv:21:**Unit tests (Soltest)**,Optimisation tests for Homestead EVM version^M
 docs/contributing/tests-list.csv:22:,Optimisation tests for Byzantium EVM version^M
 docs/contributing/tests-list.csv:23:,Optimisation tests for Constantinople EVM version^M
 docs/contributing/tests-list.csv:24:,Optimisation tests for Constantinople EVM version with ABI encoder v2 and YUL optimisation^M
 docs/contributing/tests-list.csv:25:,YUL optimisation tests for Constantinople EVM version with ABI encoder v2^M

```
Please check that your changes are working as intended.

stackenbotten commented:
There was an error when running `chk_coding_style` for commit `ccafd6a96b38d1b29dcf2399cbad797dfa04e46f`:
```
Error: Trailing whitespace found:
 docs/contributing/tests-list.csv:1:M-oM-;M-?Test suite,Tests^M
 docs/contributing/tests-list.csv:2:**Command line**,Passing unknown options to CLI^M
 docs/contributing/tests-list.csv:3:,Passing files that are not found^M
 docs/contributing/tests-list.csv:4:,Passing files that are not files^M
 docs/contributing/tests-list.csv:5:,Passing empty remappings^M
 docs/contributing/tests-list.csv:6:,General tests (`Full details <https://github.com/ethereum/solidity/tree/develop/test/cmdlineTests>`_)^M
 docs/contributing/tests-list.csv:7:**Compilation**,Compiling other contracts and libraries^M
 docs/contributing/tests-list.csv:8:,MultiSigWallet^M
 docs/contributing/tests-list.csv:9:,Corion^M
 docs/contributing/tests-list.csv:10:,Gnosis^M
 docs/contributing/tests-list.csv:11:,Milestone tracker^M
 docs/contributing/tests-list.csv:12:,String Utils^M
 docs/contributing/tests-list.csv:13:,Documentation examples^M
 docs/contributing/tests-list.csv:14:,Library checksum^M
 docs/contributing/tests-list.csv:15:,Long library names^M
 docs/contributing/tests-list.csv:16:,Linking libraries^M
 docs/contributing/tests-list.csv:17:,Overwriting already compiled files^M
 docs/contributing/tests-list.csv:18:,"Assemble, yul, strict-assembly and optimize compiled outputs"^M
 docs/contributing/tests-list.csv:19:,Standard input to compiler^M
 docs/contributing/tests-list.csv:20:,soljson via the fuzzer^M
 docs/contributing/tests-list.csv:21:**Unit tests (Soltest)**,Optimisation tests for Homestead EVM version^M
 docs/contributing/tests-list.csv:22:,Optimisation tests for Byzantium EVM version^M
 docs/contributing/tests-list.csv:23:,Optimisation tests for Constantinople EVM version^M
 docs/contributing/tests-list.csv:24:,Optimisation tests for Constantinople EVM version with ABI encoder v2 and YUL optimisation^M
 docs/contributing/tests-list.csv:25:,YUL optimisation tests for Constantinople EVM version with ABI encoder v2^M

```
Please check that your changes are working as intended.

bshastry commented:
This may need rebasing

ChrisChinchilla commented:
@bshastry Done

bshastry commented on docs/contributing.rst:
```diff
@@ -81,47 +81,73 @@ Thank you for your help!
 where ``TestName`` can be a wildcard ``*``.
 
-For example, here is an example test you might run;
+Or, for example, to run all the tests for the disambiguator:
```
```suggestion
Or, for example, to run all the tests for the yul disambiguator:
```

bshastry commented:
Could you please squash the two commits?

ChrisChinchilla commented:
@bshastry Done

bshastry commented on docs/contributing.rst:
```diff
@@ -81,47 +81,73 @@ Thank you for your help!
 
-If you do not have libz3 installed on your system, you should disable the SMT tests:
-``./scripts/soltest.sh --no-smt``.
+If the ``evmone`` library is not found, tests that use it are skipped. These tests
```
This sentence is a duplicate of `If it is not found, the relevant tests ￼are skipped.` a few lines above.

ChrisChinchilla replied:
I had a merge conflict before last push, so maybe I missed something in the tidy up

ChrisChinchilla replied:
Actually, this isn't duplication, the first paragraph says how to get evmone, the other says what tests are skipped if it can't be found. To make it clearer, I merged the paragraphs

bshastry commented on docs/contributing.rst:
```diff
@@ -81,47 +81,73 @@ Thank you for your help!
-The test system will automatically try to discover the location of the ``evmone`` library
+The test system automatically tries try to discover the location of the ``evmone`` library
 starting from the current directory. The required file is called ``libevmone.so`` on Linux systems,
 ``evmone.dll`` on Windows systems and ``libevmone.dylib`` on MacOS. If it is not found, the relevant tests
```
```suggestion
 ``evmone.dll`` on Windows systems and ``libevmone.dylib`` on MacOS. If it is not found, the following tests (part of the soltest suite) are skipped:
```

ChrisChinchilla replied:
Ahh, I see what you mean. OK, I won't accept this suggestion, but move things around a bit to reflect some of what you suggest.

bshastry replied:
Sure, it was just a suggestion :-)

bshastry commented on docs/contributing.rst:
```diff
@@ -81,47 +81,73 @@ Thank you for your help!
+The test system automatically tries try to discover the location of the ``evmone`` library
 starting from the current directory. The required file is called ``libevmone.so`` on Linux systems,
 ``evmone.dll`` on Windows systems and ``libevmone.dylib`` on MacOS. If it is not found, the relevant tests
 are skipped. To run all tests, download the library from
```
```suggestion
``libsolididty/semanticTests``, ``libsolidity/GasCosts``, and ``libsolidity/SolidityEndToEndTest``. To run all tests, download the library from
```

bshastry commented on docs/contributing.rst:
```diff
@@ -81,47 +81,73 @@ Thank you for your help!
-``./build/test/soltest --help`` has extensive help on all of the options available.
-See especially:
+If the ``libz3`` library is not installed on your system, you should disable the
+SMT tests using the ``--no-smt`` option when running ``./scripts/tests.sh`` or
```
I tested this but failed to disable SMT tests. I did

```
$ ./scripts/tests.sh --no-smt
...
test_0571770b984517916a9a795a2b5f8a37ce1c20f5eee047a30f3e5315718c32d4_s
ecurity_considerations_rst.sol
Unexpected compilation result:       
Expected failure: 0 - Expected warning / error output: 0
Was failure: 0

Warning: BMC analysis was not possible since no integrated SMT solver (
Z3 or CVC4) was found.
While calling:
...
```

Are you sure this flag works as intended?

ChrisChinchilla replied:
Adding the possibility is in this branch, did you try when checked out / built from this branch?

bshastry replied:
Yes, I remember checking out the branch before I ran the command in my previous comment in this thread.

ChrisChinchilla replied:
Hmm, odd, it works for me. You generally have to rebuild everything to get tests correct, but I don't think they are affected by this… I can check on Linux I guess, but this just copied syntax I saw elsewhere in our bash scripts

bshastry replied:
Okay, I'll check again (my system is Linux anyway).

bshastry commented on docs/contributing/tests-list.csv:
```diff
@@ -0,0 +1,26 @@
+,Optimisation tests for Constantinople EVM version
+,Optimisation tests for Constantinople EVM version with ABI encoder v2 and YUL optimisation
+,YUL optimisation tests for Constantinople EVM version with ABI encoder v2
+,Optimisation tests for St Petersburg EVM version
```
Don't we also have tests for Istanbul now?

ChrisChinchilla replied:
Quite likely, this PR has been here for a while 😆. I'll check and add if so.

bshastry commented on docs/contributing.rst:
```diff
@@ -81,47 +81,73 @@ Thank you for your help!
 
-If you do not have libz3 installed on your system, you should disable the SMT tests:
-``./scripts/soltest.sh --no-smt``.
+If the ``evmone`` library is not found, tests that use it are skipped. These tests
```
```suggestion
```

bshastry commented on docs/contributing.rst:
```diff
@@ -81,47 +81,73 @@ Thank you for your help!
-If you do not have libz3 installed on your system, you should disable the SMT tests:
-``./scripts/soltest.sh --no-smt``.
+If the ``evmone`` library is not found, tests that use it are skipped. These tests
+are ``libsolididty/semanticTests``, ``libsolidity/GasCosts``, ``libsolidity/SolidityEndToEndTest``
```
```suggestion
```

bshastry commented on docs/contributing.rst:
```diff
@@ -81,47 +81,73 @@ Thank you for your help!
-``./scripts/soltest.sh --no-smt``.
+If the ``evmone`` library is not found, tests that use it are skipped. These tests
+are ``libsolididty/semanticTests``, ``libsolidity/GasCosts``, ``libsolidity/SolidityEndToEndTest``
+part of the soltest suite.
```
```suggestion
```

bshastry commented on scripts/tests.sh:
```diff
@@ -36,6 +36,11 @@ source "${REPO_ROOT}/scripts/common.sh"
 WORKDIR=`mktemp -d`
 CMDLINE_PID=
 
+if [ "$1" = --no-smt ]
```
I think this is wrong. Can't recall how to compare bash strings properly, but if I do I'll post it here

leonardoalt replied:
Now removed. The right way is to export `SMT_FLAGS=--no-smt` before calling `tests.sh`.

bshastry reviewed (dismissed):
Added some more comments

Commit:
Update contributing doc to clarify tests

Fix line endings

Update docs/contributing.rst

Co-Authored-By: Bhargava Shastry <bhargava.shastry@ethereum.org>

Remove duplication

Add back URL to correct evmone version

Add Istanbul

ChrisChinchilla commented:
@bshastry I took your suggestions and incorporated them into an update

ChrisChinchilla commented:
Rebased

chriseth commented on docs/contributing.rst:
```diff
@@ -81,47 +81,70 @@ Thank you for your help!
-- `run_test (-t) <https://www.boost.org/doc/libs/1_69_0/libs/test/doc/html/boost_test/utf_reference/rt_param_reference/run_test.html>`_ to run specific tests cases, and
-- `report-level (-r) <https://www.boost.org/doc/libs/1_69_0/libs/test/doc/html/boost_test/utf_reference/rt_param_reference/report_level.html>`_ give a more detailed report.
+.. csv-table:: The full test suite
+   :file: ./contributing/tests-list.csv
```
I don't think this is very helpful. What is the information we convey here? This list will be outdated very soon and I don't think we want to have the overhead of updating it.

ChrisChinchilla replied:
I have no knowledge of what tests are changing and when, and this PR has sat here for a long time. I'm reacting to the Zeppelin audit and comments from contributors that they don't understand what tests we have and what they do, hence a reference might be useful to them. Otherwise there are a lot of tests that are supposed to pass before someone can make a contribution, but most people (including some core team hires) take time to understand what they are. So a reference seems helpful to all of these people. If things change, that's fine, but a reference is still useful.

If you render the RsT, or look at the csv it lists what tests run as part of which test suite and what they are testing.

I am open to other suggestions (like maybe commenting the test files instead), but as this has come up a bit, I feel it's valuable to contributors.

leonardoalt replied:
@chriseth opinion after @ChrisChinchilla 's reply?

leonardoalt replied:
@bshastry The error you pointed out was actually a bug in the SMTChecker docs, should be fixed by https://github.com/ethereum/solidity/pull/8470. The `--no-smt` flag works for `soltest` and `isoltest`, but doesn't relay anything to the command line tests.

bshastry replied:
@leonardoalt Are you referring to the comment about SMT checker not being disabled by the bash tests script (https://github.com/ethereum/solidity/pull/7632#discussion_r358373498) or the other one about string comparison in bash (https://github.com/ethereum/solidity/pull/7632#discussion_r358379555)?

leonardoalt replied:
First one

bshastry replied:
> The --no-smt flag works for soltest and isoltest, but doesn't relay anything to the command line tests.

I'm a bit confused. Do you mean `$ ./scripts/tests.sh --no-smt` will work as intended post #8470 ?

leonardoalt replied:
`--no-smt` has always worked when relayed to `soltest` and `isoltest`. What didn't work was the command line tests when no solver is available. That is fixed by #8470

bshastry replied:
Okay, thanks @leonardoalt 

In that case, please ignore my earlier comment

leonardoalt replied:
@chriseth The `csv` file was removed

leonardoalt commented on docs/contributing.rst:
```diff
@@ -81,47 +81,70 @@ Thank you for your help!
+Some tests require the `evmone <https://github.com/ethereum/evmone/releases>`_
+library, others require `libz3 <https://github.com/Z3Prover/z3>`_. The test script
+tries to discover the location of the ``evmone`` library, which can be located in
+the current directory, or the ``deps`` folder in the project top level. The required
```
```suggestion
the current directory, installed on the system level, or the ``deps`` folder in the project top level. The required
```

leonardoalt reviewed (commented):
I'm taking this over.

leonardoalt commented:
This is ready for review again

chriseth commented:
Rebased and squashed.
